library/core/src/ptr/non_null.rs

```diff
@@ -1513,10 +1515,17 @@ impl<T> NonNull<[T]> {
     ///
     /// (Note that this example artificially demonstrates a use of this method,
     /// but `let slice = NonNull::from(&x[..]);` would be a better way to write code like this.)
+    // Safety: https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html
     #[stable(feature = "nonnull_slice_from_raw_parts", since = "1.70.0")]
     #[rustc_const_stable(feature = "const_slice_from_raw_parts_mut", since = "1.83.0")]
     #[must_use]
     #[inline]
+    #[requires(data.pointer.is_aligned() 
```

Comment:
I haven't noticed previously that this function is safe. As mentioned in its documentation:
> This function is safe, but dereferencing the return value is unsafe.

In this case, we don't need any of the pre-conditions.

Reply:
@zhassan-aws does that mean we shouldn't check the byte by byte of the result since it may lead to unsafe/undefined behaviors?

Reply:
Correct. If we were to write the contract for correctness, it should specify that the resulting slice matches the provided raw parts, i.e. that the slice length is the same as the given length and the _pointer_ to the beginning of the slice is the same as the provided pointer (`data`). However, since we're focusing on the safety contracts at the moment, this is not needed.

We could do the equality comparison in the harness if the harness provides a valid pointer, but the comparison shouldn't be part of the contract.

Reply:
Got it thank you! Added the ensures clause for correctness.